mlir/test/Examples/nvgpu/tools/nvdsl.py

```diff
@@ -0,0 +1,415 @@
+
+    def __init__(
+        self,
+        shape,
+        memref_ty,
+        swizzle=nvgpu.TensorMapSwizzleKind.SWIZZLE_NONE,
+        l2promo=nvgpu.TensorMapL2PromoKind.L2PROMO_NONE,
+        oob=nvgpu.TensorMapOOBKind.OOB_ZERO,
+        interleave=nvgpu.TensorMapInterleaveKind.INTERLEAVE_NONE,
+    ):
+        self.swizzle = swizzle  # mlir.nvgpu.TensorMapSwizzleKind
+        self.l2promo = l2promo  # mlir.nvgpu.TensorMapL2PromoKind
+        self.oob = oob  # mlir.nvgpu.TensorMapOOBKind
+        self.interleave = interleave  # mlir.nvgpu.TensorMapInterleaveKind
+        self.shape = shape
+        self.memref_ty = memref_ty  # MemRefType
+        self.lastDim = 64
+        self.requiredLoad = 1
+        self.tma_shape = shape
+        self.tma_memref = ir.MemRefType.get(shape, memref_ty.element_type)
+
+    @property
+    def tensormap_descriptor_ty(self):
+        """Returns a tensormap descriptor type."""
+        memref_str = f"memref<{self.tma_shape[0]}x{self.tma_shape[1]}x{self.memref_ty.element_type}, 3>"
+        parse_str = f"!nvgpu.tensormap.descriptor<tensor = {memref_str},\
+                                              swizzle = {self.swizzle},\
+                                              l2promo = {self.l2promo},\
+                                              oob = {self.oob},\
+                                              interleave = {self.interleave}>"
+
+        return ir.Type.parse(parse_str)
+
+    def create_descriptor(self, device_ptr):
+        tma_descriptor_ty = self.tensormap_descriptor_ty
+        device_unranked_memref = memref.CastOp(
+            ir.UnrankedMemRefType.get(
+                self.memref_ty.element_type, self.memref_ty.memory_space
+            ),
+            device_ptr,
+        )
+        self.tma_descriptor = nvgpu.TmaCreateDescriptorOp(
+            tma_descriptor_ty, device_unranked_memref, map(const, self.tma_shape)
+        )
+        return self.tma_descriptor.result
+
+    def prefetch(self, predicate=None):
+        nvgpu.tma_prefetch_descriptor(self.tma_descriptor, predicate=predicate)
+
+    def load(self, dest, mbarrier: Mbarriers, coords=[0, 0], predicate=None):
+        coord_ops = [const(c) for c in coords]
+        nvgpu.TmaAsyncLoadOp(
+            dest,
+            mbarrier.mbar_group_op,
+            self.tma_descriptor,
+            coordinates=coord_ops,
+            mbarId=mbarrier.id_op,
+            predicate=predicate,
+        )
+
+
+class MatrixAccumulator:
+    def __init__(self, M, N, ty):
+        self.M = M
+        self.N = N
+        self.ty = ty
+
+    @property
+    def acc_ty(self):
+        return ir.Type.parse(
+            "!nvgpu.warpgroup.accumulator<fragmented=vector<"
+            + str(self.M)
+            + "x"
+            + str(self.N)
+            + "x"
+            + str(self.ty)
+            + ">>"
+        )
+
+    def op(self):
+        return nvgpu.warpgroup_mma_init_accumulator(self.acc_ty)
+
+
+class Matrix:
```

Comment:
Can we have one line summary describing the scope of each class. For e.g., this has the following """Create Wgmma descriptors and and execute Wgmma Hopper Tensor Core operation""". 

IMO, scope of the class should be limited with the name of the class suggesting its scope. For e.g., I would call this one `GmmaDescriptorBuilder` analogous to your `TMA` class. We are passing two instance of `Matrix` class to a function `Matrix::matmul`, which suggests that descriptor creation can be separated from Tensor Core operation. Let me know what do you think of the below organization?

```py
class GmmaDescriptorBuilder:
   def __init__(self, tma_descriptor: TMA, M, N): # no smem here as it is updated and a new desc_op is created
      ...
   
   @property
   def wgmma_ty(self)
      ...
    
   def desc_op(self, smem): 
      """Returns a new desc_op for a wgmma_ty and shared memory argument"""
        return nvgpu.warpgroup_generate_descriptor(
            self.wgmma_ty, smem, self.tma_descriptor.tma_descriptor
        )
```

The users of the DSL will be writing code as:
```py
        # 2. Performs Tensor Core GEMM 64x64x64 by warpgroup
        A = GmmaDescriptorBuilder(x_tma, N, N) # with this the line can be out of the mainloop as it just capturing the non-changing part in the GMMA descriptor?
        B = GmmaDescriptorBuilder(y_tma, N, N) # this line can be out of the mainloop as it just capturing the non-changing part in the GMMA descriptor?
        C = MatrixAccumulator(N, N, T.f32()).op()
        D = nvgpu.WarpgroupMmaOp(C.type, A.desc_op(a_smem), B.desc_op(b_smem), C, transposeB=True)

        # 3. Stores fragmented registers to global memory by warpgroup
        nvgpu.warpgroup_mma_store(D, z_dev)
```

Reply:
This is also an option.

Alternatively, we can unify `Accumulator` and `Input` matrices into same class, and overload the operators so easy usage. 

How about this? I have `WGMMAMatrix` class, it overloads `__matmul__ (@)` and `__iadd__ (+)`. 

```py
accumulator = WGMMAMatrix(WGMMAType.Accumulator, shape=[TILE_M, TILE_N], ty=T.f32())
C = accumulator.generate_accumulator()

for_op = scf.ForOp(...), [C, ...])
    with ir.InsertionPoint(for_op.body):

    ...

   # Initialize matrices
	A = WGMMAMatrix(WGMMAType.Input, [TILE_M, TILE_K], desc=x_tma, smem=x_smem)
	B = WGMMAMatrix(WGMMAType.Input, [TILE_K, TILE_N], desc=y_tma, smem=y_smem)
	C = WGMMAMatrix(WGMMAType.Accumulator, acc_op=for_op.inner_iter_args[0])

	# Matrix Multiply
	C += A @ B

	scf.yield_([C, ...])
```

Reply:
Looks better. I would name things what they are. `WGMMAType.[Desc|Input]` to have it future proof and avoid confusion when we have wgmma RS support. It will be great to separate the moving part in A and B from the part that remains constant within the mainloop. How about a slight tweak below:


```mlir
accumulator = WGMMAMatrix(WGMMAType.Accumulator, shape=[128, 128], ty=T.f32())
A = WGMMAMatrix(WGMMAType.Desc, [128, 64], desc=x_tma, smem=x_smem)
B = WGMMAMatrix(WGMMAType.Desc, [64, 128], desc=y_tma, smem=y_smem)
C = accumulator.generate_accumulator()

for_op = scf.ForOp(...), [C, ...])
    with ir.InsertionPoint(for_op.body):

    ...

   # Initialize matrices
   A.update(x_smem) 
   B.update(y_smem) 
   C = WGMMAMatrix(WGMMAType.Accumulator, acc_op=for_op.inner_iter_args[0])

   # Matrix Multiply
   C += A @ B

   scf.yield_([C, ...])
```mlir

Reply:
Nice, yes it is a great idea. I updated the PR 
[see here](https://github.com/llvm/llvm-project/pull/87065/files#diff-0d87502ea89e4e6183f1cc59816cfc0595a14591a111b59e4c0ded64daddbd3aR155-R183)

Reply:
there's a much nicer/better way to write this: 

```python

from mlir.dialects.scf import for_ as range_, yield_

K = 10
for i, C in range_(K, iter_args=[C]):
     A.update(x_smem) 
     B.update(y_smem) 
     C = WGMMAMatrix(WGMMAType.Accumulator, acc_op=C)
     ...
     yield_([C])
```

and this will indeed emit exactly the IR you're emitting above. See [test/scf.py](https://github.com/llvm/llvm-project/blob/main/mlir/test/python/dialects/scf.py#L65-L177).

Reply:
actually there's no way to get the result of the op using that builder (`for_`) but it would just making a small change so that it to also yields the result of the op [here](https://github.com/llvm/llvm-project/blob/main/mlir/python/mlir/dialects/scf.py#L135-L137).